### DISCUSSION TITLE
Question on a C struct / Pointers and modules

### DISCUSSION THREAD

**Discussion by ghost** [micropython/micropython#13767] [Libraries & Drivers]
Hi all,
I m porting some existing C code to Micropython (and im not very familiar with it). Im currently trying to find a way to exchange a data buffer between the 2 worlds (Mpy and C mod), but i have 1 big problem (or question :) )... 

From the Mpy side, i would like to fill a struct like below and send it to the C module:
```
typedef struct _cellular_t {
	const char *MyNetworkName;
	const char *myApn;
	unit16_t 	rssiLevel;
} cellular_t;
```

I wrote the following Mpy code:
```
import uctypes

c_cellular_t = {
    "netFlags"  : uctypes.UINT16,
    "myApn" : (uctypes.PTR | 0, uctypes.UINT8),
	}

# Allocate memory	
dbytes= bytearray(uctypes.sizeof(c_cellular_t, uctypes.LITTLE_ENDIAN))
# Map strust to memory
bmem =uctypes.struct(uctypes.addressof(dbytes), c_cellular_t, uctypes.LITTLE_ENDIAN)
# Fill the values
bmem.netFlags = 32
apn = b"googleiot.com"
bmem.myApn[:] = uctypes.addressof(apn)		<<= FAIL !!!
```

...and i can't assign the pointer, it fails :(

Any help will be welcome !!!

**Comment by jimmo** [comment]
@XT107234 Sorry I don't really see how the first C example maps to the structure you're defining with `uctypes`... what is netFlags?

Here is how to populate that C structure from Python:

```py
>>> # Define string values
>>> apn = b"googleiot.com"
>>> network_name = b"hello world"
>>> 
>>> # Define structure
>>> import uctypes
>>> c_cellular_t = { "MyNetworkName": 0 | uctypes.PTR, "myApn": 4 | uctypes.PTR, "rssiLevel": 8 | uctypes.UINT16 }
>>>
>>> # Create buffer and structure
>>> dbytes = bytearray(uctypes.sizeof(c_cellular_t, uctypes.LITTLE_ENDIAN))
>>> bmem = uctypes.struct(uctypes.addressof(dbytes), c_cellular_t, uctypes.LITTLE_ENDIAN)
>>>
>>> # Set fields in structure
>>> bmem.MyNetworkName = uctypes.addressof(network_name)
>>> bmem.myApn = uctypes.addressof(apn)
>>> bmem.rssiLevel = 72
>>> 
>>> # Check resulting structure layout
>>> dbytes
bytearray(b'`\xd4\x01 \xe0\xce\x01 H\x00')
>>> 
>>> # Verify struct is correct
>>> import struct
>>> apn_ptr, net_ptr, rssi = struct.unpack('<IIH', dbytes)
>>> uctypes.bytearray_at(apn_ptr, 10)
bytearray(b'hello worl')
>>> uctypes.bytearray_at(net_ptr, 10)
bytearray(b'googleiot.')
>>> rssi
72
```

**Comment by ghost** [comment]
Thanks a lot  @jimmo for the code, you can ignore netFlags (a "copy/paste" mistake).  As far I can see,  my problem came from the struct definition: _MyNetworkName": ( 0 | uctypes.PTR, , uctypes.UINT8)_ where _ uctypes.UINT8)_ was leading to the issue.

Btw, I don't understand something in your snippet:
on 
```
apn_ptr, net_ptr, rssi = struct.unpack('<IIH', dbytes)
```
Why _'<IIH'_  ...shouldn't it be something like _'<PPH'_ ?
On my system (Linux), I receive a segfault on _uctypes.bytearray_at(net_ptr, 10)_  do you have any clue ?

**Reply by jimmo** [reply]
> Btw, I don't understand something in your snippet:

What would you like to know?

**Reply by ghost** [reply]
Why '<IIH' ...shouldn't it be something like '<PPH' ?
On my system (Linux), I receive a segfault on uctypes.bytearray_at(net_ptr, 10) do you have any clue ?
(it crashes on _objstr.c_, at _mp_str_print_quoted()_)

**Reply by robert-hh** [reply]
The data size for format characters 'I' and 'P' can be different. For 'I' it is always 4, but for 'P' it's for instance 8 on my PC, and 4 on a RP2.

**Reply by ghost** [reply]
You're right @robert-hh ...and that's the problem ! It seems that **uctypes.PTR** , when defining a struct (on Linux64), stores on 4 bytes instead of 8 ...
The following code works:
```
# Define string values
apn = b"googleiot.com"
network_name = b"hello world"

# Define structure
import uctypes
c_cellular_t = { "MyNetworkName": 0 | uctypes.UINT64, "myApn": 8 | uctypes.UINT64, "rssiLevel": 16 | uctypes.UINT16 }
# Create buffer and structure
dbytes = bytearray(uctypes.sizeof(c_cellular_t, uctypes.LITTLE_ENDIAN))
bmem = uctypes.struct(uctypes.addressof(dbytes), c_cellular_t, uctypes.LITTLE_ENDIAN )

# example Set fields in structure
bmem.MyNetworkName = uctypes.addressof(network_name)
bmem.myApn = uctypes.addressof(apn)
bmem.rssiLevel = 72


# Verify struct is correct
import struct

hci_sco = uctypes.struct(
	uctypes.addressof(dbytes),
		c_cellular_t,
		uctypes.LITTLE_ENDIAN
	)

print(uctypes.bytearray_at(hci_sco.MyNetworkName, 32))
print(uctypes.bytearray_at(hci_sco.myApn, 32))
print(hci_sco.rssiLevel)
```
but changing the struct definition to a more logical (?) one :
`c_cellular_t = { "MyNetworkName": 0 | uctypes.PTR, "myApn": 8 | uctypes.PTR, "rssiLevel": 16 | uctypes.UINT16 }`
will raise a segfault.
